### PR TITLE
Improve etcd fail-over scenarios

### PR DESCRIPTION
**Pull request by tgraf**
The main change introduced by this PR Is to fail Cilium status if etcd quorum cannot be detected for 3 consecutive status intervals. By triggering a restart of the agent, the situation can hopefully be resolved. In addition, the endpoint list is shuffled on bootstrap to further improve the chances to resolve the issue.

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

File: pkg/kvstore/etcd_test.go [hunk 1]
```diff
@@ -1642,3 +1642,23 @@ func TestGetSvcNamespace(t *testing.T) {
...
+	}
+
+	same := 0
+	for i := range s1 {
+		if s1[i] == s2[i] {
+			same++
+		}
+	}
+	if same == len(s1) {
+		t.Errorf("Shuffle() did not modify s2")
```
Comment (qmonnet) [inline pkg/kvstore/etcd_test.go]: Your string slice has only three elements, so the number of possible permutations is very low. Unless I'm missing something, this test function will have `s1 == s2` (and make the test fail) once in every six runs on average. You might want a longer string and/or running this a couple of times in a loop to fix that?

File: pkg/rand/safe_rand_test.go [hunk 2]
```diff
@@ -34,3 +34,28 @@ func TestIntn(t *testing.T) {
...
+	}
+
+	same := 0
+	for i := range s1 {
+		if s1[i] == s2[i] {
+			same++
+		}
+	}
+	if same == len(s1) {
+		t.Errorf("Shuffle() did not modify s2")
```
Comment (qmonnet) [inline pkg/rand/safe_rand_test.go]: Same remark.

**Review by qmonnet — CHANGES_REQUESTED** [review]
One concern on the tests, looks good to me otherwise.

**Review by aanm — APPROVED** [review]

File: pkg/kvstore/etcd_test.go (hunk 1 above)
Comment (tgraf) [reply pkg/kvstore/etcd_test.go]: Yes, I was able to reproduce. I actually had to use more than 3 entries as with 3 entries, `Shuffle()` generated the same sequence (still differing from the original sequence) even with differing seed. Given that 3 is a very common number of etcd endpoints, I'm looking into a better way to Shuffle.

**Review by qmonnet — APPROVED** [review]
Looks good!

**Comment by tgraf** [comment]
test-me-please

**Review by brb — APPROVED** [review]

**Pull request merged**